lib/lightning_web/live/components/oauth.ex

```diff
@@ -143,200 +157,418 @@ defmodule LightningWeb.Components.Oauth do
+        }
+
+      :missing_expiration ->
+        {
+          "Invalid Token Response",
+          :reauthorize,
+          "The OAuth token from #{provider || "the provider"} is missing expiration information."
+        }
+
+      _ ->
+        {
+          "OAuth Error",
+          :generic,
+          build_generic_message(provider)
+        }
+    end
+  end
+
+  @spec handle_missing_scopes_error(map() | nil, String.t() | nil) ::
+          {String.t(), atom(), String.t()}
+  defp handle_missing_scopes_error(error_details, provider) do
+    missing_scopes = get_in(error_details, [:missing_scopes]) || []
+    missing_count = length(missing_scopes)
+
+    case missing_count do
+      0 ->
+        {
+          "Missing Required Permissions",
+          :reauthorize,
+          "Some required permissions were not granted. Please ensure you grant all selected permissions when authorizing with #{provider || "your provider"}."
+        }
+
+      1 ->
+        scope = List.first(missing_scopes)
+
+        {
+          "Missing Required Permission",
+          :reauthorize,
+          "The '#{scope}' permission was not granted. Please ensure you select this permission when authorizing with #{provider || "your provider"}."
+        }
+
+      _ ->
+        scope_list = Enum.join(missing_scopes, "', '")
+
+        {
+          "Missing Required Permissions",
+          :reauthorize,
+          "The following permissions were not granted: '#{scope_list}'. Please ensure you select all required permissions when authorizing with #{provider || "your provider"}."
+        }
+    end
+  end
+
+  @spec handle_missing_refresh_token_error(map() | nil, String.t() | nil) ::
+          {String.t(), atom(), String.t()}
+  defp handle_missing_refresh_token_error(error_details, provider) do
+    existing_available =
+      get_in(error_details, [:existing_token_available]) || false
+
+    if existing_available do
+      {
+        "Missing Refresh Token",
+        :use_existing,
+        "We didn't receive a refresh token from #{provider || "this provider"}. You may have already granted access to OpenFn via another credential. If you have another credential, please use that one."
+      }
+    else
+      {
+        "Missing Refresh Token",
+        :reauthorize,
+        "Please reauthorize to provide OpenFn with the necessary refresh token for #{provider || "your provider"}."
+      }
+    end
+  end
+
+  @spec build_generic_message(String.t() | nil) :: String.t()
+  defp build_generic_message(provider) do
+    "Please try authorizing with #{provider || "your provider"} again. If the problem persists, contact support."
+  end
+
+  @spec determine_action_button(map(), atom()) :: map()
+  defp determine_action_button(assigns, suggested_action) do
+    case suggested_action do
+      action when action in [:reauthorize, :generic] ->
+        if assigns[:revocation_endpoint] do
+          %{
```

Comment:
is it possible to have a struct to have these typed?

Reply:
Yes, bro, thanks for the catch.